### PR TITLE
Add curve, drop support to python 2 and add support to interfaces

### PR DESCRIPTION
**Pull request by lincrosenbach**

### CODE REVIEW

File: src/barril/units/unit_database.py [hunk 1]
```diff
@@ -1190,7 +1181,7 @@ def Convert(self, category_or_quantity_type, from_unit, to_unit, value):
         this = self.GetInfo(quantity_type, from_unit, fix_unknown=True)
         other = self.GetInfo(quantity_type, to_unit, fix_unknown=True)
 
-        if isinstance(value, (float,) + six.integer_types):  # , numpy.ndarray)):
+        if isinstance(value, (float,) + (int,)):  # , numpy.ndarray)):
```
Comment (nicoddemus) [inline src/barril/units/unit_database.py]:
```suggestion
        if isinstance(value, (float, int)):
```

File: src/barril/units/unit_database.py [hunk 2]
```diff
@@ -1396,7 +1387,7 @@ def _DoOperationResultingInNewQuantity(
 # ===================================================================================================
```
Comment (nicoddemus) [inline src/barril/units/unit_database.py]: Remove header (please do a find/replace to remove them all)

File: tox.ini [hunk 3]
```diff
@@ -1,17 +1,16 @@
 [tox]
-envlist = py27, py36, linting, docs
+envlist = py36, linting, docs
```
Comment (nicoddemus) [inline tox.ini]:
```suggestion
envlist = py36, py37, linting, docs
```

File: tox.ini [hunk 4]
```diff
@@ -1,17 +1,16 @@
...
 
 [travis]
 python =
     3.6: py36
```
Comment (nicoddemus) [inline tox.ini]:
```suggestion
     3.6: py36
     3.7: py37
```

**Comment by nicoddemus** [comment]
Good work! I wen't into nitpick mode here 😁

**Comment by nicoddemus** [comment]
Please update AppVeyor to remove `py27`.

File: src/barril/_foundation/types_.py [hunk 5]
```diff
@@ -23,7 +19,7 @@ def _GetKnownNumberTypes():
     performance penalties.
     """
     result = {float, complex}
```
[the remark this answers is not in the document]
Comment (lincrosenbach) [reply src/barril/_foundation/types_.py]: Done

File: src/barril/_foundation/types_.py [hunk 6]
```diff
@@ -373,7 +369,7 @@ def IterFlattened(iterable, skip_types=None):
     if skip_types is None:
         skip_types = []
 
-    for s in six.string_types:
+    for s in (str,):
         if s not in skip_types:
```
[the remark this answers is not in the document]
Comment (lincrosenbach) [reply src/barril/_foundation/types_.py]: Done

File: src/barril/curve/__init__.py [hunk 7]
```diff
@@ -0,0 +1,4 @@
+
+# Basic extensions to the python language or standard library.
```
[the remark this answers is not in the document]
Comment (lincrosenbach) [reply src/barril/curve/__init__.py]: Removed

File: src/barril/curve/curve_interface.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
+
+from oop_ext.interface import Interface
+
+
+# ===================================================================================================
```
[the remark this answers is not in the document]
Comment (lincrosenbach) [reply src/barril/curve/curve_interface.py]: Removed

File: src/barril/units/__init__.py [hunk 9]
```diff
@@ -100,6 +106,11 @@
     str("FixedArray"),  # pylint: disable=invalid-all-object
     str("FractionScalar"),  # pylint: disable=invalid-all-object
     str("GetUnknownQuantity"),  # pylint: disable=invalid-all-object
+    str("IArray"),  # pylint: disable=invalid-all-object
```
[the remark this answers is not in the document]
Comment (lincrosenbach) [reply src/barril/units/__init__.py]: Removed

File: src/barril/units/_quantity.py [hunk 10]
```diff
@@ -879,7 +866,7 @@ def __ne__(self, other):
         return not self == other
 
     # Used to cast a constant value
-    BASE_NUMBER = six.integer_types + (float,)
+    BASE_NUMBER = (int,) + (float,)
```
[the remark this answers is not in the document]
Comment (lincrosenbach) [reply src/barril/units/_quantity.py]: Done

File: src/barril/units/unit_database.py (hunk 2 above)
Comment (lincrosenbach) [reply src/barril/units/unit_database.py]: Done

File: src/barril/units/_definitions.py [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+    "IQuantity",
+    "IScalar",
+    "IQuantity2",
+    "IQuantity3",
+    "IObjectWithQuantity",
+    "IArray",
+]
+
+
+# ===================================================================================================
```
[the remark this answers is not in the document]
Comment (lincrosenbach) [reply src/barril/units/_definitions.py]: Done

File: src/barril/units/_definitions.py [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+
+    def GetUnknownCaption(self):
+        """
+        :rtype: unicode
+        :returns:
+            The caption that's set to be shown to the user when it's an unknown unit.
+        """
+
+
+# ===================================================================================================
```
[the remark this answers is not in the document]
Comment (lincrosenbach) [reply src/barril/units/_definitions.py]: Done.

File: src/barril/units/_definitions.py [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+
+    def IsValid(self):
+        """
+        :rtype: bool
+        :returns:
+            True if the current value is valid, False otherwise.
+        """
+
+
+# ===================================================================================================
```
[the remark this answers is not in the document]
Comment (lincrosenbach) [reply src/barril/units/_definitions.py]: Done.

File: src/barril/units/_scalar.py [hunk 14]
```diff
@@ -199,7 +193,7 @@ def SetFormattedValueFormat(cls, value_format):
         try:
             value_format % 1.11
         except TypeError as e:
-            Reraise(
+            raise (
                 e,
                 "Incompatible format for Scalar value. Expected a format for a float value.",
             )
```
[the remark this answers is not in the document]
Comment (lincrosenbach) [reply src/barril/units/_scalar.py]: Done

File: src/barril/units/_tests/test_posc2.py [hunk 15]
```diff
@@ -34,21 +31,19 @@ def testMegagramPerCubicMeterToKilogramPerCubicMeter():
...
-    ):
+    for category, category_info in unit_database.categories_to_quantity_types.items():
         base_unit = category_info.default_unit
         for unit in unit_database.GetValidUnits(category):
             for i in [-1, 0, 1]:
                 try:
                     unit_database.Convert(category, base_unit, unit, i)
                 except Exception as e:
-                    Reraise(e, "Error converting: from: %s to: %s" % (base_unit, unit))
+                    raise (e, "Error converting: from: %s to: %s" % (base_unit, unit))
```
[the remark this answers is not in the document]
Comment (lincrosenbach) [reply src/barril/units/_tests/test_posc2.py]: Done

File: src/barril/units/_tests/test_posc2.py [hunk 16]
```diff
@@ -34,21 +31,19 @@ def testMegagramPerCubicMeterToKilogramPerCubicMeter():
...
 
                 try:
                     unit_database.Convert(category, unit, base_unit, i)
                 except Exception as e:
-                    Reraise(e, "Error converting: from: %s to: %s" % (unit, base_unit))
+                    raise (e, "Error converting: from: %s to: %s" % (unit, base_unit))
```
[the remark this answers is not in the document]
Comment (lincrosenbach) [reply src/barril/units/_tests/test_posc2.py]: Done

File: src/barril/units/unit_database.py [hunk 17]
```diff
@@ -654,18 +647,16 @@ def CheckCategoryUnit(self, category, unit):
         :raises InvalidUnitError:
             if the unit provided is not accepted for this category
         """
-        assert category.__class__ == six.text_type, "Expected unicode. Found: %s" % (
-            category,
-        )
-        assert unit.__class__ == six.text_type, "Expected unicode. Found: %s" % (unit,)
+        assert category.__class__ == str, "Expected unicode. Found: %s" % (category,)
+        assert unit.__class__ == str, "Expected unicode. Found: %s" % (unit,)
```
[the remark this answers is not in the document]
Comment (lincrosenbach) [reply src/barril/units/unit_database.py]: Done.

File: src/barril/units/unit_database.py (hunk 1 above)
Comment (lincrosenbach) [reply src/barril/units/unit_database.py]: Done

File: tox.ini (hunk 4 above)
Comment (lincrosenbach) [reply tox.ini]: Added

File: tox.ini (hunk 3 above)
Comment (lincrosenbach) [reply tox.ini]: Added

File: src/barril/units/unit_database.py [hunk 18]
```diff
@@ -654,18 +647,16 @@ def CheckCategoryUnit(self, category, unit):
         :raises InvalidUnitError:
             if the unit provided is not accepted for this category
         """
-        assert category.__class__ == six.text_type, "Expected unicode. Found: %s" % (
-            category,
-        )
-        assert unit.__class__ == six.text_type, "Expected unicode. Found: %s" % (unit,)
+        assert category.__class__ == str, "Expected unicode. Found: %s" % (category,)
```
[the remark this answers is not in the document]
Comment (lincrosenbach) [reply src/barril/units/unit_database.py]: Done.

File: src/barril/units/_tests/test_operations.py [hunk 19]
```diff
@@ -21,7 +19,7 @@ def testOperation():
         operator.mul,
         operator.mod,
     ]
-    SCALAR_TYPES = six.integer_types + (float,)
+    SCALAR_TYPES = (int,) + (float,)
```
[the remark this answers is not in the document]
Comment (lincrosenbach) [reply src/barril/units/_tests/test_operations.py]: Done.

File: src/barril/units/_definitions.py [hunk 20]
```diff
@@ -0,0 +1,255 @@
...
+        """
+        :rtype: IQuantity
+        :returns:
+            The quantity that is associated with this object.
+            The quantity may be writable or read-only, depending on the object.
+            Actually, it may also implement some of the other IQuantity* interfaces.
+        """
+
+
+# ===================================================================================================
```
[the remark this answers is not in the document]
Comment (lincrosenbach) [reply src/barril/units/_definitions.py]: Done.

File: src/barril/units/_definitions.py [hunk 21]
```diff
@@ -0,0 +1,255 @@
...
+            A special use-case is provided if there's a single category with an exponent 1, in which
+            case a string is returned (optimization).
+            I.e.: instead of returning tuple('length',), 'length' will be returned.
+
+        .. see:: GetComposingUnits to return the actual units/exponents for the categories
+        """
+
+    def GetComposingUnits(self):
+        """
+        :rtype: tuple(tuple(unicode, int)) or unicode
```
[the remark this answers is not in the document]
Comment (lincrosenbach) [reply src/barril/units/_definitions.py]: Done

File: src/barril/units/_definitions.py [hunk 22]
```diff
@@ -0,0 +1,255 @@
+
```
[the remark this answers is not in the document]
Comment (lincrosenbach) [reply src/barril/units/_definitions.py]: Renamed

File: src/barril/units/_definitions.py [hunk 23]
```diff
@@ -0,0 +1,255 @@
...
+from oop_ext.interface import Interface
+
+"""
+The basic interfaces for the coilib50.units module should be defined here
```
[the remark this answers is not in the document]
Comment (lincrosenbach) [reply src/barril/units/_definitions.py]: Done.

File: src/barril/units/_definitions.py [hunk 24]
```diff
@@ -0,0 +1,255 @@
+
+from oop_ext.interface import Interface
```
[the remark this answers is not in the document]
Comment (lincrosenbach) [reply src/barril/units/_definitions.py]: Done

File: src/barril/_foundation/klass.py [hunk 25]
```diff
@@ -48,20 +39,17 @@ def IsInstance(p_object, p_class_name):
...
-# IsSubclass
-# ===================================================================================================
 def IsSubclass(p_class, p_class_name):
     """
     Like the built-in issubclass, but also accepts a class name as parameter.
     """
     is_instance = isinstance  # put it in locals
 
-    if is_instance(p_class_name, six.string_types):
+    if is_instance(p_class_name, (str,)):
```
Comment (nicoddemus) [inline src/barril/_foundation/klass.py]:
```suggestion
    if is_instance(p_class_name, str):
```

File: src/barril/_foundation/klass.py [hunk 26]
```diff
@@ -74,7 +62,7 @@ def IsSubclass(p_class, p_class_name):
     elif (
         is_instance(p_class_name, tuple)
         and len(p_class_name) > 0
-        and is_instance(p_class_name[0], six.string_types)
+        and is_instance(p_class_name[0], (str,))
```
Comment (nicoddemus) [inline src/barril/_foundation/klass.py]:
```suggestion
        and is_instance(p_class_name[0], str)
```

File: src/barril/_foundation/types_.py [hunk 27]
```diff
@@ -233,13 +204,10 @@ def IsBasicType(value, accept_compound=False, additional=None):
 
 
 _ACCEPTED_BASIC_TYPES = tuple(
-    list(six.string_types) + list(six.integer_types) + [bytes, float, bool, complex]
+    list((str,)) + list((int,)) + [bytes, float, bool, complex]
```
Comment (nicoddemus) [inline src/barril/_foundation/types_.py]:
```suggestion
    [str, int, bytes, float, bool, complex]
```

File: src/barril/basic/fraction/_fraction.py [hunk 28]
```diff
@@ -1,18 +1,15 @@
...
 
 from fractions import Fraction as StdFraction
 from functools import total_ordering
 
-import six
-
 SMALL = 1e-8
 
-NumberType = (six.integer_types, float)
+NumberType = ((int,), float)
```
Comment (nicoddemus) [inline src/barril/basic/fraction/_fraction.py]:
```suggestion
NumberType = (int, float)
```

File: src/barril/units/_tests/test_scalar.py [hunk 29]
```diff
@@ -1,15 +1,11 @@
 """
-    REMARKS: Unicode
-        The "unit" must not containt unicode character. All unit name and representation must use
-        the standard ascii representation.
+    The 'unit' must not containt str character. All unit name and representation must use
```
Comment (nicoddemus) [inline src/barril/units/_tests/test_scalar.py]:
```suggestion
    'unit' must be ascii only.
```

And delete the line below

**Comment by nicoddemus** [comment]
Test coverage indicates `ChangeScalars` is not tested, can you please add a quick test for it?

File: src/barril/_foundation/klass.py (hunk 25 above)
Comment (lincrosenbach) [reply src/barril/_foundation/klass.py]: Done

File: src/barril/_foundation/klass.py (hunk 26 above)
Comment (lincrosenbach) [reply src/barril/_foundation/klass.py]: Done

File: src/barril/basic/fraction/_fraction.py (hunk 28 above)
Comment (lincrosenbach) [reply src/barril/basic/fraction/_fraction.py]: Done

File: src/barril/units/_tests/test_scalar.py (hunk 29 above)
Comment (lincrosenbach) [reply src/barril/units/_tests/test_scalar.py]: Done

File: src/barril/_foundation/types_.py (hunk 27 above)
Comment (lincrosenbach) [reply src/barril/_foundation/types_.py]: Done

**Comment by lincrosenbach** [comment]
> Test coverage indicates `ChangeScalars` is not tested, can you please add a quick test for it?

Sure

**Pull request merged**